### PR TITLE
Release 2.4

### PR DESCRIPTION
**Pull request by DavertMik**
## Motivation/Description of the PR

Applicable helpers:

- [ ] Webdriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe

- Description of this PR, which problem it solves
- A link to the corresponding issue (if applicable).

## Type of change

- [ ] Breaking changes
- [ ] New functionality
- [ ] Bug fix
- [ ] Documentation changes/updates
- [ ] Hot fix
- [ ] Markdown files fix - not related to source code

## Checklist:

- [ ] Tests have been added
- [ ] Documentation has been added (Run `npm run docs`)
- [ ] Lint checking (Run `npm run lint`)
- [ ] Local tests are passed (Run `npm test`)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@ testpullfilecache*
 package-lock.json
 yarn.lock
 /.vs
+typings/types.d.ts
```
Comment (DavertMik) [inline .gitignore]: I added this to ignore conflicts in our PRs

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,27 @@
+## 2.4.0
+
+* Improved setup wizard with `npx codecept init`:
+  * **enabled [retryFailedStep](/plugins/#retryfailedstep) plugin for new setups**.
+  * enabled [@codeceptjs/configure](/configuration/#common-configuration-patterns) to toggle headless/window mode via env variable
+  * creates a new test on init
+  * removed question on "steps file", create it by default.
+* Added [pauseOnFail plugin](/plugins/#pauseonfail). *Sponsored by Paul Vincent Beigang and his book "[Practical End 2 End Testing with CodeceptJS](https://leanpub.com/codeceptjs/)"*.
+* Added [`run-rerun` command](/commands/#run-rerun) to run tests multiple times to detect and fix flaky tests. By @Ilrilan and @Vorobeyko.
+* Added [`Scenario.todo()` to declare tests as pending](/basics#todo). See []()
```
Comment (Vorobeyko) [inline CHANGELOG.md]: see empty array?)

File: docs/basics.md [hunk 3]
```diff
@@ -539,6 +544,47 @@ Feature('Complex JS Stuff').retry(3);
 Every Scenario inside this feature will be rerun 3 times.
 You can make an exception for a specific scenario by passing the `retries` option to a Scenario.
 
+### Rerunning Flaky Tests Multiple Times
```
Comment (pablopaul) [inline docs/basics.md]: @DavertMik I would remove "Rerunning Flaky Tests Multiple Times" from "basics.md" and just leave it in "advanced.md", currently it seems to be duplicated to me

File: docs/basics.md [hunk 4]
```diff
@@ -460,7 +460,7 @@ This is an **essential feature to debug flaky tests**, as you can analyze them i
 > ℹ To enable pause after a test without a plugin use `After(pause)` inside a test file.
 
 
-### Screenshot on Failure
+### Screenshot on Failure <Badge text="Enabled by default"/>
```
Comment (elukoyanov) [inline docs/basics.md]: Do we need to add this badge to all enabled by default plugins?

What about disabled?

File: docs/basics.md [hunk 5]
```diff
@@ -471,8 +471,16 @@ To see how the test was executed, use [stepByStepReport Plugin](/plugins/#stepby
 
 ## Retries
 
+### Auto Retry
+
+You can auto-retry a failed step by enabling [retryFailedStep Plugin](/plugins/#retryfailedstep).
+
+> **autoRetry plugin is enabled by default** for new setups since CodeceptJS 2.4
```
Comment (elukoyanov) [inline docs/basics.md]: Why this differs from 463, 453 rows?

Do we need badges or quoted messages?

File: docs/basics.md [hunk 6]
```diff
@@ -759,12 +810,14 @@ Like in Mocha you can use `x` and `only` to skip tests or to run a single test.
 * `xScenario` - skips current test
 * `Scenario.only` - executes only the current test
 
-## Todo test
+## Todo
```
Comment (elukoyanov) [inline docs/basics.md]: Why `Todo test` was removed?

IMO, `Todo` looks like we forgot to add docs :)
How about `Not implemented tests (todo)`

File: docs/commands.md [hunk 7]
```diff
@@ -90,24 +90,33 @@ npx codeceptjs run-workers 3
...
+```
+
+For this command configuration is required:
+
+```js
 {
-...
-rerun: {
-  //how many times all tests in suite must pass 
+  // inside to codecept.conf.js
```
Comment (elukoyanov) [inline docs/commands.md]: "inside to" ?
maybe simple "inside"?

File: docs/helpers/Appium.md [hunk 8]
```diff
@@ -121,161 +114,199 @@ let browser = this.helpers['Appium'].browser
 
 -   `config`  
 
-### _switchToContext
+### runOnIOS
```
Comment (elukoyanov) [inline docs/helpers/Appium.md]: Is there any reason to remove alphabetical order  from doc?

File: docs/helpers/FileSystem.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ---
-permalink: helpers/FileSystem
-editLink: https://github.com/Codeception/CodeceptJS/blob/master/lib/helper/FileSystem.js
```
Comment (elukoyanov) [inline docs/helpers/FileSystem.md]: Is there any reason to remove these links?

File: .gitignore (hunk 1 above)
Comment (elukoyanov) [reply .gitignore]: Should `types.d.ts` be deleted?

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,27 @@
...
+* Added support for absolute path for `output` dir. See #2049 by @elukoyanov
+* Fixed error in `npxcodecept init` caused by calling `console.print`. See #2071 by @Atinux.
```
Comment (elukoyanov) [inline CHANGELOG.md]: `npxcodecept` without spaces?

File: .gitignore (hunk 1 above)
Comment (DavertMik) [reply .gitignore]: Probably not, to accidentally not to release a package without types.

File: CHANGELOG.md (hunk 2 above)
Comment (DavertMik) [reply CHANGELOG.md]: Thanks! It was 5 am, I didn't notice )

File: CHANGELOG.md (hunk 10 above)
Comment (DavertMik) [reply CHANGELOG.md]: yeah, 5 am...

File: .gitignore (hunk 1 above)
Comment (elukoyanov) [reply .gitignore]: And types will stay outdated in code

File: package.json [hunk 11]
```diff
@@ -1,10 +1,15 @@
...
+  "version": "2.4.0",
+  "description": "Supercharged End 2 End Testing Framework for NodeJS",
   "keywords": [
     "acceptance",
     "end2end",
+    "end 2 end",
+    "puppeteer",
+    "webdriver",
+    "testcafe",
+    "protractor",
```
Comment (PeterNgTr) [inline package.json]: how about adding `api tests` cause we also support it.

File: docs/helpers/Appium.md (hunk 8 above)
Comment (DavertMik) [reply docs/helpers/Appium.md]: I had to write a custom method aggregation (see runio.js) changes to merge methods from WebDriver and Appium classes. This was part of this.

File: docs/basics.md (hunk 4 above)
Comment (DavertMik) [reply docs/basics.md]: Removed the badge

File: docs/basics.md (hunk 5 above)
Comment (DavertMik) [reply docs/basics.md]: Yeah, added quotes

File: docs/basics.md (hunk 3 above)
Comment (DavertMik) [reply docs/basics.md]: Yeah, I put this twice but this is not basics information

File: docs/helpers/FileSystem.md (hunk 9 above)
Comment (DavertMik) [reply docs/helpers/FileSystem.md]: This was not working in Vuepress :(

File: .gitignore (hunk 1 above)
Comment (DavertMik) [reply .gitignore]: Ah, you mean remove from git? Ok

**Comment by DavertMik** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/rerun.js  4
         

Clones added
============
- lib/rerun.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/DavertMik/CodeceptJS/pullRequest?prid=4795387&bid=15962168)

**Comment by DavertMik** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/rerun.js  4
         

Clones added
============
- lib/rerun.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/DavertMik/CodeceptJS/pullRequest?prid=4795387&bid=15962168)

**Pull request merged**